feat: `Joints` implementation

Neosoulink opened the pull request:
## Summary

This PR introduces the [Joints](https://rapier.rs/docs/user_guides/javascript/joints/) feature.

Joints are one of the advanced features natively available in `Rapier`.
This PR provides an easy, basic but robust way to deal with it.

### Logs

- Support for Joint types:
  - `"fixed" | "spring" | "rope" | "generic" | "spherical" | "prismatic" | "revolute"`
- New components based on the supported joint types (`FixedJoint`, `SphericalJoint`, ...)

### How to use

Here's a basic implementation example:
```vue
<RigidBody
  v-for="(ref, i) in bodyRefs"
  :key="i"
  :ref="ref"
  :type="i === 0 ? 'kinematic' : 'dynamic'"
  :position="[i * 1.5, 0, 0]"
   collider="ball"
 >
  <TresMesh>
    <TresSphereGeometry />
    <TresMeshNormalMaterial />
  </TresMesh>
</RigidBody>

<SphericalJoint
  v-for="(ref, i) in bodyRefs"
  :key="i"
  :bodies="[ref.value?.[0]?.instance, bodyRefs[i - 1]?.value?.[0]?.instance]"
  :params="[
    [-1.1, 0, 0],
    [1.1, 0, 0],
  ]"
/>
```
> Preview

https://github.com/user-attachments/assets/e30dc274-597f-4903-bcc6-4060711e5494

> Will close #108

Neosoulink committed:
feat(core): base `joint` integration

- Init joint component
  - Add `JointProps` interface
- Add Joint Demo

Neosoulink committed:
feat(core): add `rope` & `spherical` joint types

Neosoulink committed:
feat(core): add common object types

Neosoulink committed:
feat(playground): implement spherical joint demo

Neosoulink committed:
feat(core): implement all the supported `Joint`s

Neosoulink committed:
style(core): use `spaces`

Neosoulink committed:
docs: add `Joint` base documentation

Neosoulink commented:
Hi @alvarosabu, I just added a base documentation for `Joint`.
I'd appreciate having your review of it (the `Joint` implementation in general).

cc: @JaimeTorrealba

Neosoulink committed:
chore(setting): minor correct

Neosoulink commented:
Any update on this one?

> This PR is waiting for review

JaimeTorrealba commented on .vscode/settings.json:
```diff
@@ -53,5 +53,10 @@
-  }
+  },
+
+  "cSpell.words": [
```
I think the settings should be removed here, this is personal to our environments.

Neosoulink replied:
got it ✅

JaimeTorrealba commented on docs/components/joint.md:
```diff
@@ -0,0 +1,66 @@
+
+Joints is an extension feature provided in [Rapier#Joint](https://rapier.rs/docs/user_guides/javascript/joints/). It lets us connect two or more bodies, restricting their movements according to each other.
+
+In **Tres** we can achieve such motion restriction by using one of the available components designed to handle joints:
```
Small one but maybe we need consistency here it all pages should be "Tres.js"

Neosoulink replied:
Noted

JaimeTorrealba commented on docs/components/joint.md:
```diff
@@ -0,0 +1,66 @@
+
+<template>
+  <RigidBody
+    v-for="(ref, i) in bodyRefs"
```
I got a problem, the mental load of this example is high. I would instead, in this page, come with a very basic example, just two rigid Bodies joined (just that). The idea here is that they learn how to use it.

Neosoulink replied:
This example reflects the one in the playground (the `JointsDemo` example)...
Is it okay to simplify this one in this documentation but keep the demo as it is?

JaimeTorrealba commented on playground/src/pages/basics/JointsDemo.vue:
```diff
@@ -0,0 +1,74 @@
+  Array.from({ length: 10 }).map(() => shallowRef<ExposedRigidBody>(null)),
+)
+
+setInterval(() => {
```
Why setInterval?

And if you decided to use setInterval, please remember to clear the interval in the onUnmouted

Neosoulink replied:
Oh yeah, you're right,

`setInterval` was used to update the rotation of the first and have that whip effect

JaimeTorrealba reviewed (changes requested):
I like the implementation, But also I would add a composable WDYT? With the same logic

an `useJoint()` or similar so the people is not force to use it with Tres.js components only

JaimeTorrealba commented on src/components/joints/BaseJoint.vue:
```diff
@@ -0,0 +1,185 @@
+  }
+
+  if (hasParamsError) {
+    throw new Error(`Invalid "${type}" joint parameters`)
```
Maybe too difficult, but could be said what parameters have problems?

Neosoulink replied:
Sure, maybe a string containing the requirement or why it the error occurs

Neosoulink commented:
> I like the implementation, But also I would add a composable WDYT? With the same logic
> 
> an `useJoint()` or similar so the people is not force to use it with Tres.js components only

I think it possible even to have both features

JaimeTorrealba commented:
> > I like the implementation, But also I would add a composable WDYT? With the same logic
> > an `useJoint()` or similar so the people is not force to use it with Tres.js components only
> 
> I think it possible even to have both features

Completely, maybe another PR to be more organized.

Neosoulink committed:
refactor: apply @JaimeTorrealba request changes

### Description

- Make the Joint component listen to updates
- Provide more precise `Joint` error messages
- Simplify the Joint example and the Joint documentation
- Remove `cSpell.words` VsCode setting

Neosoulink commented:
> Completely, maybe another PR to be more organized.

I pushed the changes requested, and yes, this will be part of a separate PR ✅

#### Here what has changed:
- [c1a6ffd1769b6a872c48a42bcd54bf02333658e1](https://github.com/Tresjs/rapier/pull/150/commits/c1a6ffd1769b6a872c48a42bcd54bf02333658e1)
  - Make the Joint component listen to updates
  - Provide more precise `Joint` error messages
  - Simplify the Joint example and the Joint documentation
  - Remove `cSpell.words` VsCode setting

alvarosabu committed:
Merge branch 'main' into feat/joints

Neosoulink commented on docs/components/joint.md:
````diff
@@ -15,25 +15,33 @@ All of them extends [BaseJoint](../../src/components/joints/index.ts#L96).
 
 ```vue
-<script setup lang="ts">
+<script setup>
````
Got it

Neosoulink commented on docs/components/joint.md:
Noted

Neosoulink commented on docs/components/joint.md:
```diff
@@ -43,23 +51,21 @@ const bodyRefs = shallowRef(
 
 > Preview
-<img width="843" alt="Screenshot 2024-12-09 at 4 01 58PM" src="https://github.com/user-attachments/assets/9baaea58-f996-45d5-b4f4-34dae30d44cb">
+<img width="1141" alt="JointsDemo" src="https://github.com/user-attachments/assets/d3cacac3-8764-4906-886a-d0b7a764b7c0" />
```
It was automatically added from github, but let make it clean. I'm going to update it

Neosoulink commented on src/components/joints/index.ts:
This approach priorities re-usability based on the same component.
In other words, it's just a shortcut for the `Joints`

It's exactly the same as doing:

```tsx
<Joint type="joint-type" />
```

> It's a pattern I often used to implement a few years ago back in 2019 (old goods days 🥲)
> It's a pattern I discovered while learning how Vue was working under the hood

Neosoulink replied:
Hey!
While updating it, I realized I did the same for the `Colliders`, Why not make the change on a separate PR, to update the pattern? (for `Joints` & `Colliders`)

Neosoulink committed:
refactor(docs): apply changes request from #150

alvarosabu committed:
Merge branch 'main' into feat/joints
